### ISSUE TITLE
Integrate NDimArray and Matrix

### ISSUE THREAD

**Issue by JSS95** [sympy/sympy#20570]
Matrix is two-dimensional array, which is a special (and important) case of N-dimensional array. Mathematically, these two should be fully compatible but current SymPy does not support this.

SymPy's matrix is very well developed. It supports implicit-element matrix such as `MatrixSymbol` and element indexing by `MatrixElement`. However `NDimArray` does not support these. I am planning to clean up the class inheritance structure and add these features to `NDimArray`.

@Upabjojr 's [comment](https://github.com/sympy/sympy/pull/20072#issuecomment-691951014) describes current types:

> * classes inheriting `MatrixCommon`: explicit-valued matrices (all elements of the matrix are expressed),
> * classes inheriting `MatrixExpr` but not `MatrixCommon`: implicit-valued matrices (i.e. the matrix is just a symbol, no elements are shown),
> * `MatrixElement` class: I'd call it implicit-valued (as elements are not shown), explicit-indexed (as indices are shown, unlike `MatrixExpr`),
> * classes inheriting `NDimArray`: like explicit matrices, but with N-axes (matrices always have 2 axes),
> * `IndexedBase` and `Indexed` objects: generic indexed objects, implicit-valued explicit-index,
> * classes in `sympy.codegen.array_utils`: implicit-valued implicit-indexed arrays, still experimental.
> * `sympy.tensor.tensor`: abstract index notation for `tensorial` expressions. This is reserved for applications in theoretical physics.

I am planning to introduce classes such as `ArrayCommon`, `ArrayExpr` and let matrix classes subclass their counterpart of `Array` classes. I will appreciate any suggestion!

**Comment by oscarbenjamin** [comment]
What are the benefits of making this change?

**Comment by Upabjojr** [comment]
> SymPy's matrix is very well developed. It supports implicit-element matrix such as `MatrixSymbol` and element indexing by `MatrixElement`. However `NDimArray` does not support these. I am planning to clean up the class inheritance structure and add these features to `NDimArray`.

Have a look at the code in `sympy.codegen.array_utils`. That code should probably be moved to `sympy/tensor/array`. `_CodegenArrayAbstarct` could potentially be renamed `ArrayExpr`.

**Comment by JSS95** [comment]
@oscarbenjamin 

In physics, 1d and 2d array are very commonly used when tensor notation is unnecessary (especially when the coordinate system is Cartesian). For example normal vector is expressed by 1d array, stress tensor is by 2d array, and the traction is the matrix multiplication between these two. We can construct the general, unevaluated notation of this with `MatrixSymbol` and `MatMul` .

But higher order tensors cannot be combined to this system since classes such as `ArraySymbol` do not exist. For example, in rheology the 2-dimensional viscous stress tensor is double dot product of 4-dimensional viscosity tensor and 2-dimensional strain rate tensor. There is no way to construct this using current SymPy's `NDimArray`. We need classes for array expression for this.

Also, we need to allow the operation between `Matrix` and `NDimArray` and let it return a coherent result. Outer product between two 2-d matrices should return 4-d array, and derivative of 1-d matrix by 1-d matrix should return 2-d matrix. For this the two classes should be integrated, and I think the best way to do this is to make `Matrix` as subclass of `NDimArray`.

**Comment by Upabjojr** [comment]
> But higher order tensors cannot be combined to this system since classes such as `ArraySymbol` do not exist.

In a certain (vague) sense, `IndexedBase` could be thought of a simplified version of an array symbol. I agree that a dedicated `ArraySymbol` class would be better.

**Comment by JSS95** [comment]
@Upabjojr What about making `MatrixSymbol` a subclass of `IndexedBase`, and `MatrixElement` a subclass of `Indexed`?

**Comment by Upabjojr** [comment]
> @Upabjojr What about making `MatrixSymbol` a subclass of `IndexedBase`, and `MatrixElement` a subclass of `Indexed`?

No, their behaviour and use cases are too different.

**Comment by sylee957** [comment]
I'm also aware of lots of inconveniences handling vector in matrix context, because there is no 1-dimensional array representing it.
Would this be able to fix this?

**Comment by JSS95** [comment]
@sylee957 Can you give example?

**Comment by sylee957** [comment]
Differential for row vector with row vector would give `(n, 1, n, 1)` which is different from `(n, n)` jacobian. 
The other problem is we need to define dot product between row vector and column vector.
And we also arrive whether to define `(n, 1, 1)` array as a vector.
I think that the matrix framework should better begun with objects that were compatible with n-dimensional arrays because vector dot product or vector differential actually crosses over between n-dimensional array and matrix. This should be fixed simultaneously in the most backward compatible ways.

**Comment by JSS95** [comment]
I see. I think this can be fixed by converting row and column vector to 1-dim array when needed. For example, dot product of two vectors `a` and `b` can be represented in three different ways:
1. Matrix multiplication.
Here, `a` should be the row vector and `b` should be the column vector.
2. Dot product.
Here, row vector and column vector need not be distinguished. If `a` and `b` were row and column vector, they can be just converted to 1-d array.
3. Tensor product and contraction
Same with dot product.

Differential for `(n,1)` vector should be same. It first needs to be converted to 1-d `(n,)` array, and then differentiated.

**Comment by sylee957** [comment]
The only nondegenerate way of converting n-dimensional vector like array into 1-dimensional vector in mathematical notation is using [vectorization](https://en.wikipedia.org/wiki/Vectorization_(mathematics)) as far as I'm aware, so I think that everything which had assumed row or column vector before can be put the artificial constrict like that to embed into 1-dimensional vector. I think that there is a need to introduce symbolic class for vectorization, but this can be better decided after whether or not to merge matrix expression with array expression.

**Comment by Upabjojr** [comment]
> For example, dot product of two vectors `a` and `b` can be represented in three different ways

We already support all three ways to represent the dot product. We need more integration in terms of conversion of one representation to another.

**Comment by Upabjojr** [comment]
> I think that there is a need to introduce symbolic class for vectorization, 

Technically, `sympy.tensor.array` objects already support `.reshape( )` method. I guess if you call `array_instance.reshape(-1)` you get the vectorized array form.

There is currently no way to represent the reshaping operator in unevaluated (i.e. `ArrayExpr`-equivalent) form.

> but this can be better decided after whether or not to merge matrix expression with array expression.

No, please don't merge matrix and array expressions. It will be a nightmare to keep all code compatibility.

**Comment by sidhu1012** [comment]
I think `Arrays` should be made one to one with `numpy` arrays and we should make `Array` and `Matrix` operations constant which currently give `NotImplemented` in most of the cases.

The best way I could think is to introduce a new `MatArray` class to handle `Matrix` and `Array` combinations.

**Comment by JSS95** [comment]
@ Upabjojr
> We already support all three ways to represent the dot product.

Yes, but `DotProduct` is defined in `matrices/expressions` and I'm not sure if it can take arrays as argument. I believe `DotProduct` can be moved to new `array/expressions` module. Also there is no class to represent unevaluated tensor contraction which I wish to implement.

**Comment by sylee957** [comment]
I'm not sure how the compatibility was the the necessary outcome or it was from the bad design choice at the starting point.
And the bad thing about this is we need to implement methods for matrix and array operations separately, although some may be made to share the common algorithm (transpose vs permutedims, matmul vs tensorcontraction).
But if things have been designed through this way, we should  make matrix operations more closed under matrix context, such as vectorization can be defined as embedding to column vector, and all the embeddings of matrix into higher dimensional arrays can be made with kronecker product instead of tensor product.

**Comment by asmeurer** [comment]
I agree with @Upabjojr that it's better to keep matrices and arrays separate for now. If it makes sense in the future, we can merge them, but it would be better to develop the array objects out first. That way, we don't have to worry about breaking the matrices as we develop arrays, and it will be much clearer how to merge the two if it makes sense (it also may turn out to not make sense, since matrices have a lot of functionality that doesn't generalize to n-dimensional tensors).

And +1 for being able to represent 1-dimensional vectors independently of matrices. It is better to be completely explicit about the number of dimensions and not implicitly drop 1s from shapes. This is more mathematically correct and avoids bugs that could be caught by shape mismatch errors. Also it's important for code generation, since not all targets work like C where `a[0]` is the same as `a[0][0]`.

**Comment by asmeurer** [comment]
With that being said, we should definitely mimic the MatrixExpr class design with ArrayExpr, particularly the way MatrixExpr expressions don't do any evaluation by default. This is a good class design that we should try to do more of in SymPy in general.

**Comment by JSS95** [comment]
Okay, then I will focus on `ArrayExpr` by now.

**Comment by Upabjojr** [comment]
> I believe `DotProduct` can be moved to new `array/expressions` module. 

There are `tensorcontraction` and `tensorproduct` in `sympy/tensor/array`. If you combine the two of them properly, you can get a dot product.

> Also there is no class to represent unevaluated tensor contraction which I wish to implement.

Yes, there is. Look up `CodegenArrayContraction` in `sympy.codegen.array_utils`.

**Comment by JSS95** [comment]
I will move these classes from `codegen` to `array`.

**Comment by Upabjojr** [comment]
> I will move these classes from `codegen` to `array`.

I think it's also time to rename them. I created that module as a temporary support for the codegen initially... but then it developed into a proper expression-tree module for N-dim arrays.

**Comment by JSS95** [comment]
Of course. I will make enhancement if possible.

> There are tensorcontraction and tensorproduct in sympy/tensor/array. If you combine the two of them properly, you can get a dot product.

Yes, but I think dot product between tensors is so widely used that it deserves a separate class. Also, double dot product (or more) is very common in continuum mechanics so I think it would be awesome to allow n-dot product with this class.

**Comment by Upabjojr** [comment]
> Yes, but I think dot product between tensors is so widely used that it deserves a separate class. Also, double dot product (or more) is very common in continuum mechanics so I think it would be awesome to allow n-dot product with this class.

I'm not sure... there isn't any straightforward way to generalize the dot product to N-dim arrays.

Maybe what is needed is an optimized way to perform tensor product and then a contraction... Calling `tensorcontraction(tensorproduct( ... ), ... )` creates first a huge N-dim array, then contracts the axes. Maybe we need on optimized function that contracts the axes before computing the tensor product.

**Comment by JSS95** [comment]
[Here](https://en.wikipedia.org/wiki/Dyadics#Product_of_dyadic_and_dyadic) is the generalized definition of dot product and double dot product for dyads.

**Comment by Upabjojr** [comment]
> [Here](https://en.wikipedia.org/wiki/Dyadics#Product_of_dyadic_and_dyadic) is the generalized definition of dot product and double dot product for dyads.

We have a module for vector calculus. Maybe it's more appropriate there.

**Comment by JSS95** [comment]
@Upabjojr 
Is vector module compatible with N-dimensional array? Does it support higher order tensor (order 4 or even more)? I want double dot product between two abstract tensors (without element), and later substitute them with concrete tensors (with element). Is is possible with vector module?

**Comment by Upabjojr** [comment]
> Is vector module compatible with N-dimensional array?

Well, it's mostly done for classical mechanics and electromagnetism. So I suppose no.

> Does it support higher order tensor (order 4 or even more)? 

No.

> I want operation between two abstract tensors (without element), and later substitute them with concrete tensors (with element) and evaluate. Is is possible with vector module?

I think so. If it isn't, it should be fixed. The N-dim array module is indeed a lot more powerful even though with a ruder syntax.

**Comment by sylee957** [comment]
It could be of a different discussion, but I think that vector and diffgeom module should be conceptually similar and they should be merged to achieve more generality

**Comment by Upabjojr** [comment]
> that vector and diffgeom module should be conceptually similar and they should be merged to achieve more generality

They are quite different. Also their usages are quite different.

**Comment by JSS95** [comment]
I took a look at vector module. It seems that the module has `Vector` and `Dyad` as separate classes. I'd say this design is very restricting. Vector is rank 1 tensor and dyad is rank 2 tensor, but mathematically there is no limit to the rank that a tensor can have. But with current design, we need to define each class for each rank manually which makes it not expandable at all. `NDimArray` does not have this problem because no matter how high the rank of array is, a single class can handle that instance.

**Comment by Upabjojr** [comment]
> It seems that the module has `Vector` and `Dyad` as separate classes. I'd say this design is very restricting. Vector is rank 1 tensor and dyad is rank 2 tensor, but mathematically there is no limit to the rank that a tensor can have. But with current design, we need to define each class for each rank manually which makes it not expandable at all.

The current design is meant to be used with rank 1 and rank 2 objects. It's true that from a generic algebraic point of view the vector module is limited, but that's not the point. The vector module is meant to create objects as used in classical mechanics. If users are not satisfied with `vector`, then they should use `sympy.tensor.array` instead.

**Comment by JSS95** [comment]
Yes, and therefore I think that `sympy.tensor.array` module should support `ArrayExpr` and various operator classes such as `DotProduct`, `CrossProduct`, etc.

**Comment by JSS95** [comment]
What I ultimately want to do is to let single `DotProduct` or `CrossProduct` class to accept both array and vector as valid argument. I am trying to introduce the system for such behavior in #20549 .

**Comment by Upabjojr** [comment]
> Yes, and therefore I think that `sympy.tensor.array` module should support `ArrayExpr` and various operator classes such as `DotProduct`, `CrossProduct`, etc.

What about leave the implementation of dot-product and cross-product in the _vector_ module and then add `_eval_rewrite_as_Array` methods in those classes?

Then you should be able to call `vector_object.rewrite(Array)` and get the equivalent expression in the other module (i.e. easy conversion of objects defined in different modules).

**Comment by Upabjojr** [comment]
> What I ultimately want to do is to let single `DotProduct` or `CrossProduct` class to accept both array and vector as valid argument. I am trying to introduce the system for such behavior in #20549 .

I see. Well, it may be interesting if this idea of introducing _kind_ actually works.

**Comment by oscarbenjamin** [comment]
> What I ultimately want to do is to let single `DotProduct` or `CrossProduct` class to accept both array and vector as valid argument.

I don't think that makes sense. An array is just an array of coefficients which can be used to represent a vector with respect to some basis (where the basis is a set of vectors). However a vector as an object is not itself tied to any particular basis. It is common in many contexts to conflate vectors with 1D arrays by considering some implied global basis. That makes it possible to overlook this distinction but the implementation of these in sympy makes it very clear that arrays and vectors are very different things.

> I am trying to introduce the system for such behavior in #20549 .

We should try to use #20549 to represent the state of the codebase as it currently is rather than making big changes or merging types together. Any future overhaul of vectors, arrays, matrices etc can also update the relevant parts of the kind system at the same time. Right now vectors and arrays are not interchangeable anywhere in sympy so they shouldn't have the same kind. Any function that wants to accept both can always just check for both:
```python
def f(arg):
    if type(arg) in (VectorKind, ArrayKind):
        ...
```
